### PR TITLE
✨ Initial commit for koa-metrics

### PR DESCRIPTION
**Pull request by tzvipm**
Closes https://github.com/Shopify/quilt/issues/110.

### CODE REVIEW

File: packages/koa-metrics/README.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# `@shopify/koa-metrics`
+
+Performance metrics implemented with DogStatsD (<https://docs.datadoghq.com/developers/dogstatsd/>) are used widely across services at Shopify. The standard metrics should be documented and following strong conventions.
```
Comment (TheMallen) [inline packages/koa-metrics/README.md]: I don't think we want to talk about Shopify here or 'should's, we should instead say what the package offers, so maybe something like 

'Opinionated performance metric tracking for Koa, implemented with DogStatsD'

File: packages/koa-metrics/src/test/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+describe('koa-metrics', () => {
+  it('works', () => {
```
Comment (TheMallen) [inline packages/koa-metrics/src/test/index.test.ts]: I don't like leaving this empty D:

I'd really like to see us at least leave empty its defining the tests we intend to write if we're going to do them in a subsequent PR.

File: packages/koa-metrics/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  host: string;
+}
+
+export default function metrics({prefix, host}: Options) {
+  const statsdHost = host.split(':')[0];
+  const statsdPort = parseInt(host.split(':')[1], 10);
+
+  return async function statsdMiddleware(ctx: Context, next: Function) {
+    const requestStartTime = process.hrtime();
+    // @ts-ignore the types here are correct. TS complains about method
```
Comment (TheMallen) [inline packages/koa-metrics/src/index.ts]: What method is potentially undefined? How can it be true that the types are correct and also we need this @ts-ignore?

Can w add a fix for this type definition to our `config/typescript` and open a PR on definitely typed instead?

File: packages/koa-metrics/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      const responseClient = client.childClient({
+        globalTags: {
+          response_code: `${ctx.response.status}`,
+          // example: 2xx, 3xx, ...
+          response_type: `${Math.floor(ctx.response.status / 100)}xx`,
+        },
+      });
+
+      const duration = process.hrtime(requestStartTime);
+      const ms = duration[0] * 1000 + duration[1] / 1e6;
```
Comment (TheMallen) [inline packages/koa-metrics/src/index.ts]: nit: 
This seems a little hard to follow without context, I think it reads a lot nicer to use some consts here to make things more readable.
```js
const timeSince = process.hrtime(requestStartTime);
const [seconds, nanoseconds]  = timeSince;

const milliseconds = 
  seconds * MILLIS_PER_SECOND + nanoseconds / NANOS_PER_MILLI
```

File: packages/koa-metrics/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      responseClient.timing('request_time', Math.round(ms));
+
+      responseClient.histogram('request_content_length', ctx.response.length);
+
+      // @ts-ignore According to the hot-shots documentation, callback
```
Comment (TheMallen) [inline packages/koa-metrics/src/index.ts]: Same comment here, I'd rather we fix this type locally and open a PR like we do in other places in our codebases than use `@ts-ignore`.

File: packages/koa-metrics/README.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+## Usage
+
+```javascript
+import Koa from 'koa';
+import metrics from '@shopify/koa-metrics';
+
+const app = new Koa();
+
+app.use(metrics(options));
```
Comment (lemonmade) [inline packages/koa-metrics/README.md]: I always find it annoying when a usage example does not actually work as-is (options not defined)

File: packages/koa-metrics/README.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+```
+
+### Options
+
+The passed in `options` object adheres to the following API:
+
+```javascript
```
Comment (lemonmade) [inline packages/koa-metrics/README.md]: Don't need this part.

File: packages/koa-metrics/README.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+The global StatsD metric name prefix; should be provided in `PascalCase`.
+
+#### `host` (required)
+
+The url for the StatsD host; should be provided in the format: `hostname:port`.
+
+## Metrics
+
+We rely on the [hot-shots](https://github.com/brightcove/hot-shots) npm package to send the metrics.
```
Comment (lemonmade) [inline packages/koa-metrics/README.md]: Do they need to know this?

File: packages/koa-metrics/README.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+The global StatsD metric name prefix is provided through the [options object](#options).
+
+This package automatically provides performance metrics for HTTP requests:
```
Comment (lemonmade) [inline packages/koa-metrics/README.md]: Weird to have a colon followed by a heading

File: packages/koa-metrics/README.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+### Standard tags
+
+* `path`
+* `request_method`
+* `response_code`
+* `response_type`
```
Comment (lemonmade) [inline packages/koa-metrics/README.md]: I think this could use a little description

File: packages/koa-metrics/README.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+### `request_queuing_time`
+
+Time before a request actually started being processed.
+
+This metric is emitted when the application start processing a request. It relies on the presence of a header `X-Request-Start` set by the first HTTP hop.
+
+### `request_content_length`
+
+This metric is based on the response header `Content-Length`. Some responses don't provide this header (chunked encoding), in this case, this will be deduced from `ctx.body` when possible, or else `undefined`.
```
Comment (lemonmade) [inline packages/koa-metrics/README.md]: Mentioned in slack, I would not recommend measuring the body as a fallback

File: packages/koa-metrics/documentation/decisions/01 - We use hot-shots as our dogstatsd client.md [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+Among those, there are 2 libraries listed under `Node.js` with support for `DogStatsD`:
+
+* `hot-shots`
+* `node-dogstatsd`
+
+`hot-shots` appears to be the clear choice between these 2 for the following reasons:
+
+1.  security - It is actively maintained
+2.  reliability - It has a test suite
+3.  performance - It has performance tests
```
Comment (lemonmade) [inline packages/koa-metrics/documentation/decisions/01 - We use hot-shots as our dogstatsd client.md]: Love that you added this 👍

File: packages/koa-metrics/package.json [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  },
+  "author": "Shopify Inc.",
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Shopify/quilt.git"
+  },
+  "bugs": {
+    "url": "https://github.com/shopify/quilt/issues"
+  },
+  "homepage": "https://github.com/shopify/quilt/packages/koa-metrics/README.md",
```
Comment (lemonmade) [inline packages/koa-metrics/package.json]: This (and all the other ones) are not actual URLs

File: packages/koa-metrics/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    ctx.state.statsd = client;
+
+    const requestQueuingTime = ctx.request.get('X-Request-Start');
+    if (requestQueuingTime) {
+      try {
+        const timeAsInt = parseInt(requestQueuingTime, 10);
+        // @ts-ignore see description of tags in the hot-shots documentation
+        client.timing('request_queuing_time', timeAsInt, {});
+      } catch (err) {
+        // ignore errors here
```
Comment (lemonmade) [inline packages/koa-metrics/src/index.ts]: Don't describe what ("ignoring errors"), describe why

File: packages/koa-metrics/src/index.ts (hunk 3 above)
Comment (lemonmade) [reply packages/koa-metrics/src/index.ts]: I don't get it either, I'd rather one cast than losing strong typing on the entire statement

File: packages/koa-metrics/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+      port: statsdPort,
+      prefix,
+      globalTags: {
+        path: ctx.path,
+        request_method: ctx.request.method,
+      },
+    });
+
+    // allow later middleware to add metrics
+    ctx.state.statsd = client;
```
Comment (lemonmade) [inline packages/koa-metrics/src/index.ts]: It kind of bugs me when we do so much in each middleware. Could we break this up into the three/ four discrete things it is doing (set up client, timing for queuing, request details, maybe split between timing and status)

File: packages/koa-metrics/src/index.ts (hunk 5 above)
Comment (lemonmade) [reply packages/koa-metrics/src/index.ts]: Agreed

File: packages/koa-metrics/README.md (hunk 8 above)
Comment (tzvipm) [reply packages/koa-metrics/README.md]: I think so, since the `hot-shots` package provides all the mechanisms for custom timing, etc.

File: packages/koa-metrics/README.md (hunk 7 above)
Comment (tzvipm) [reply packages/koa-metrics/README.md]: I do for my editor to do proper code highlighting

File: packages/koa-metrics/src/index.ts (hunk 15 above)
Comment (tzvipm) [reply packages/koa-metrics/src/index.ts]: I'd rather the users don't need to call `app.use` on 3/4 different middlewares just for metrics. Any idea what a suitable middleground might be? I'm thinking something like what we have now, but we can add a `skipDefaultMetrics` option?

Comment (lemonmade) [reply packages/koa-metrics/src/index.ts]: You can have one middleware that just composes a bunch of functions together

Comment (tzvipm) [reply packages/koa-metrics/src/index.ts]: To confirm, then, you're talking more about code style, rather than the middleware's functionality?

Comment (lemonmade) [reply packages/koa-metrics/src/index.ts]: Yes

File: packages/koa-metrics/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    // allow later middleware to add metrics
+    ctx.state.statsd = client;
+
+    try {
+      await next();
+    } finally {
+      const responseClient = client.childClient({
+        globalTags: tagsForResponse(ctx),
```
Comment (tzvipm) [inline packages/koa-metrics/src/index.ts]: I wonder if these tags should be somehow added to metrics logged within the `await next()` call? thoughts @lemonmade ? I could perhaps make a wrapper around the statsd interface that looks at ctx?

Comment (lemonmade) [reply packages/koa-metrics/src/index.ts]: I don't understand how you think that would be possible since these tags are probably not set until the very end of the middleware chain for the most part

File: packages/koa-metrics/src/tags.ts [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+  return {
+    path: ctx.path,
+    request_method: ctx.request.method,
+  };
+}
+
+export function tagsForResponse(ctx: Context) {
+  return {
+    response_code: `${ctx.response.status}`,
+    response_type: `${Math.floor(ctx.response.status / 100)}xx`,
```
Comment (lemonmade) [inline packages/koa-metrics/src/tags.ts]: Can we have an enum for the different tags you are adding?

File: packages/koa-metrics/src/index.ts [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+  return async function statsdMiddleware(ctx: Context, next: Function) {
+    const client = new StatsD({
+      host: statsdHost,
+      port: statsdPort,
+      prefix,
+      globalTags: tagsForRequest(ctx),
+    });
+
+    let timer;
```
Comment (lemonmade) [inline packages/koa-metrics/src/index.ts]: 👎 implicit `any`

File: packages/koa-metrics/src/timing.ts [hunk 19]
```diff
@@ -0,0 +1,25 @@
+import {StatsD} from 'hot-shots';
+import {Context} from 'koa';
+
+export function initTimer(client: StatsD, ctx: Context) {
```
Comment (lemonmade) [inline packages/koa-metrics/src/timing.ts]: Can you explain how this would be used for timing other than the full response time? Would it be?

File: packages/koa-metrics/src/timing.ts [hunk 20]
```diff
@@ -0,0 +1,29 @@
+import {StatsD} from 'hot-shots';
+import {Context} from 'koa';
+
+export interface Timer {
+  close(responseClient?: StatsD): void;
+}
+
+export function initTimer(client: StatsD, ctx: Context): Timer {
```
Comment (TheMallen) [inline packages/koa-metrics/src/timing.ts]: While this factory is pretty nice compared to before, I wonder if there are a couple separate concerns here? 
- a generic `Timer` that tells us the high res time in milliseconds
- getting the queue time from the header
- sending the queue time to statsD 
- sending the end time to statsD 

I picture something a bit simpler than this that has the same api

```ts
  () => {
    stop(): number
  }
```

and then the logic to send the request start and end as a separate call

and then you would do

```ts
const requestQueueTime = ctx.request.get(Headers.QueueTime);
client.timing(Tags.QueueTime, requestQueueTime);

const timer = initTimer()

// ...rest of thing

const duration = timer.stop();
responseClient.timing(Tags.EndTime, duration)
```

I feel like this split makes things a bit clearer (and means once we have our own logger that supports timers we can pull timer out and use that).

This would also make it easier for other things in the app to pull in the timer logic and use it to time whatever they care about.

File: packages/koa-metrics/src/tags.ts [hunk 21]
```diff
@@ -0,0 +1,22 @@
+import {Context} from 'koa';
+
+export enum Tags {
+  Path = 'path',
+  RequestMethod = 'request_method',
+  ResponseCode = 'response_code',
+  ResponseType = 'response_type',
```
Comment (TheMallen) [inline packages/koa-metrics/src/tags.ts]: shouldn't this enum include the queue time and such?

File: packages/koa-metrics/src/index.ts [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+
+export {Tags} from './tags';
+
+export interface Options {
+  prefix: string;
+  host: string;
+  skipInstrumentation?: boolean;
+}
+
+export default function metrics({
```
Comment (TheMallen) [inline packages/koa-metrics/src/index.ts]: should we maybe take a basic `logger` interface (defaulting to console.log) and also output the stats there?

File: packages/koa-metrics/src/timing.ts [hunk 23]
```diff
@@ -0,0 +1,29 @@
+import {StatsD} from 'hot-shots';
+import {Context} from 'koa';
+
+export interface Timer {
+  close(responseClient?: StatsD): void;
```
Comment (TheMallen) [inline packages/koa-metrics/src/timing.ts]: I think close is a bit weird of a name. If I was going to time something on a stopwatch or my phone I would probably say `end` or `stop` instead of `close` to describe finishing timing.

File: packages/koa-metrics/src/tags.ts (hunk 21 above)
Comment (tzvipm) [reply packages/koa-metrics/src/tags.ts]: Those are metrics, whereas these are tags. The idea is that each metric is allowed to have tags associated with it.

File: packages/koa-metrics/src/index.ts (hunk 22 above)
Comment (tzvipm) [reply packages/koa-metrics/src/index.ts]: I think we should do that for development, yeah.

File: packages/koa-metrics/src/Metrics.ts [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  public closeClient() {
+    // @ts-ignore According to the hot-shots documentation, callback
+    // is not required.
+    this.rootClient.close();
+  }
+
+  /* eslint-disable no-undefined */
+  private parseArgs(args: any[]): [number | undefined, Tags | undefined] {
```
Comment (tzvipm) [inline packages/koa-metrics/src/Metrics.ts]: I'm open to a better approach to parsing these args. The issue is that I like the `hot-shots` API, which allows for 2 optional arguments, and I need to parse them and then pass them through tot he underlying `hot-shots` and `log` APIs.

File: packages/koa-metrics/src/Metrics.ts [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  public histogram(name: string, value: number);
+  public histogram(name: string, value: number, sampleRate: number);
+  public histogram(name: string, value: number, tags: Tags);
+  public histogram(name: string, value: number, sampleRate: number, tags: Tags);
+  public histogram(name: string, value: number, ...args) {
+    const [sampleRate, tags] = this.parseArgs(args);
+    this.log(MetricType.Histogram, name, value, sampleRate, tags);
+    this.client.timing(name, value, sampleRate, tags);
```
Comment (tzvipm) [inline packages/koa-metrics/src/Metrics.ts]: timing -> histogram

File: packages/koa-metrics/src/Metrics.ts (hunk 24 above)
Comment (tzvipm) [reply packages/koa-metrics/src/Metrics.ts]: follow-up question: do we even need to support a `sampleRate`? Maybe I should just :fire: this and only support tags?

Comment (TheMallen) [reply packages/koa-metrics/src/Metrics.ts]: Not sure if we need to support sample rate. If not we should burn it.

File: packages/koa-metrics/src/Metrics.ts [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  public closeClient() {
+    // @ts-ignore According to the hot-shots documentation, callback
+    // is not required.
+    this.rootClient.close();
+  }
+
+  /* eslint-disable no-undefined */
+  private parseArgs(args: any[]): [number | undefined, Tags | undefined] {
```
Comment (TheMallen) [inline packages/koa-metrics/src/Metrics.ts]: Are we able to use method overloading and [tuples](https://www.typescriptlang.org/docs/handbook/basic-types.html#tuple) to type this better?

File: packages/koa-metrics/src/test/Metrics.test.ts [hunk 27]
```diff
@@ -0,0 +1,33 @@
+describe('Metrics', () => {
```
Comment (TheMallen) [inline packages/koa-metrics/src/test/Metrics.test.ts]: Let's write these tests :)

File: packages/koa-metrics/src/test/index.test.ts [hunk 28]
```diff
@@ -0,0 +1,63 @@
+describe('koa-metrics', () => {
+  it.skip('provides a Metrics client as ctx.metrics');
+
+  it.skip('uses the provided prefix for metrics metrics');
```
Comment (TheMallen) [inline packages/koa-metrics/src/test/index.test.ts]: metrics metrics?

File: packages/koa-metrics/src/test/Metrics.test.ts (hunk 27 above)
Comment (tzvipm) [reply packages/koa-metrics/src/test/Metrics.test.ts]: done :)

File: packages/koa-metrics/src/test/Metrics.test.ts [hunk 29]
```diff
@@ -0,0 +1,165 @@
...
+    });
+
+    it('logs tags with timing metrics to the logger in development', () => {
+      withEnv('development', () => {
+        const logger = jest.fn();
+        const metrics = new Metrics(defaultOptions, logger);
+        metrics.timing('foo', 123, {tag: 'value'});
+
+        expect(logger).toHaveBeenCalledTimes(1);
+        expect(logger).toHaveBeenCalledWith('timing foo:123 #tag:value');
```
Comment (tzvipm) [inline packages/koa-metrics/src/test/Metrics.test.ts]: should we be including the global tags in these logs?

**Comment by tzvipm** [comment]
@TheMallen this should be ready for another review now. Note, also, the use of distributions for `measure` instead of `histogram`. This is our recommended way to send dogstatsd metrics, although it is currently only supported via a private beta with DataDog.

File: packages/koa-metrics/src/test/Metrics.test.ts [hunk 30]
```diff
@@ -0,0 +1,165 @@
...
+      const closeFn = stats.close;
+      expect(closeFn).toHaveBeenCalledTimes(1);
+    });
+  });
+
+  describe('Timer', () => {
+    it('initTimer returns a Timer whose stop method returns an integer', () => {
+      const originalHrTime = process.hrtime;
+      process.hrtime = () => {
+        return [3, 18000000];
```
Comment (TheMallen) [inline packages/koa-metrics/src/test/Metrics.test.ts]: Can we use some consts for this to make it a bit more readable?

File: packages/koa-metrics/src/test/Metrics.test.ts [hunk 31]
```diff
@@ -0,0 +1,165 @@
...
+      };
+
+      const metrics = new Metrics(defaultOptions);
+      const timer = metrics.initTimer();
+      const elapsedTime = timer.stop();
+
+      expect(elapsedTime).toBe(3018);
```
Comment (TheMallen) [inline packages/koa-metrics/src/test/Metrics.test.ts]: can we derive this value rather than have the magic number?

File: packages/koa-metrics/src/test/index.test.ts [hunk 32]
```diff
@@ -0,0 +1,296 @@
...
+    it('tags the initial metrics client with the path', async () => {
+      const metricsMiddleware = metrics(defaultOptions);
+      const ctx = createMockContext({
+        url: '/some/path/',
+      });
+
+      await metricsMiddleware(ctx, () => {});
+      expect(MetricsMock.mock.calls[0][0]).toMatchObject({
+        globalTags: {
+          [Tags.Path]: '/some/path/',
```
Comment (TheMallen) [inline packages/koa-metrics/src/test/index.test.ts]: can we get this value from a const?

File: packages/koa-metrics/src/test/index.test.ts [hunk 33]
```diff
@@ -0,0 +1,296 @@
...
+      const ctx = createMockContext({
+        headers: {
+          'X-Request-Start': '100',
+        },
+      });
+
+      await metricsMiddleware(ctx, () => {
+        expect(MetricsMock.mock.instances[0].timing).toHaveBeenCalledWith(
+          CustomMetrics.QueuingTime,
+          100,
```
Comment (TheMallen) [inline packages/koa-metrics/src/test/index.test.ts]: same here

File: packages/koa-metrics/src/test/Metrics.test.ts (hunk 29 above)
Comment (TheMallen) [reply packages/koa-metrics/src/test/Metrics.test.ts]: I don't *think* so

File: packages/koa-metrics/src/test/index.test.ts (hunk 33 above)
Comment (tzvipm) [reply packages/koa-metrics/src/test/index.test.ts]: oops; missed this one
